Add DynamicProvisioningScheduling support for EBS

ddebroy opened the pull request:
**What this PR does / why we need it**:

This PR adds support for the DynamicProvisioningScheduling feature in EBS. With this in place, if VolumeBindingMode: WaitForFirstConsumer is specified in a EBS storageclass and DynamicProvisioningScheduling is enabled, EBS provisioner will use the selected node's LabelZoneFailureDomain as the zone to provision the EBS volume in.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Related to #63232

Sample `describe pv` output with NodeAffinity populated:

```
~$ kubectl describe pv pvc-f9d2138b-7e3e-11e8-a4ea-064124617820
Name:              pvc-f9d2138b-7e3e-11e8-a4ea-064124617820
Labels:            failure-domain.beta.kubernetes.io/region=us-west-2
                   failure-domain.beta.kubernetes.io/zone=us-west-2a
Annotations:       kubernetes.io/createdby=aws-ebs-dynamic-provisioner
                   pv.kubernetes.io/bound-by-controller=yes
                   pv.kubernetes.io/provisioned-by=kubernetes.io/aws-ebs
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:      slow3
Status:            Bound
Claim:             default/pvc3
Reclaim Policy:    Delete
Access Modes:      RWO
Capacity:          6Gi
Node Affinity:     
  Required Terms:  
    Term 0:        failure-domain.beta.kubernetes.io/zone in [us-west-2a]
                   failure-domain.beta.kubernetes.io/region in [us-west-2]
Message:           
Source:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   aws://us-west-2a/vol-0fc1cdae7d10860f6
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
Events:         <none>
```

**Release note**:
```release-note
none
```

/sig storage
/assign @msau42 @jsafrane

msau42 commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -129,6 +132,23 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
+		}
+
+		if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
+			return "", 0, nil, "", fmt.Errorf("zone/zones cannot be specified with DynamicProvisioningScheduling enabled")
```
Maybe you can suggest to use allowedTopologies instead

jsafrane replied:
or `zone/zones cannot be specified with VolumeBindingMode = WaitForFirstConsumer`.

ddebroy replied:
@msau42 , from what I understand, with `DynamicProvisioningScheduling` enabled, `AllowedTopologies` will only be relevant for scenarios like GCE where multi-zonal volumes are supported, correct? In case of EBS, I am ignoring `AllowedTopologies`. I will update the error msg to what Jan suggested above.

msau42 replied:
No, AllowedTopologies is relevant to single-zone volumes too.  It basically replaces the "zone" and "zones" parameters, and allows an admin to restrict the zones where volumes can be provisioned in.  The difference is the scheduler will respect it when choosing a node.

msau42 replied:
Here's what I think the provision implementation of single-zone volumes should do:
1. node not set, allowed topologies not set: old behavior
2. node not set, allowed topologies set: choose from allowed topologies
3. node set, zone/zones set: error always?
4. node set, allowed topologies set or not set: use node's zone
5. zone/zones set + allowedtopologies set: error

msau42 reviewed (commented):
Can you also add an e2e test for this? Under test/e2e/storage.  You can probably reuse parts of the volume_provisioning.go tests, but you'll need to modify the StorageClass and create a pod too.

New multi-zonal e2es can be considered as a separate PR too.

msau42 commented:
/ok-to-test

msau42 commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
+		} else {
+			// node not specified so collect all specified zones from allowedTopologies
+			allowedZones := make([]string, 0)
+			for _, term := range allowedTopologies {
```
This could be a volume util that gce pd could use too

msau42 replied:
Actually @verult may have some comments here since he is looking at doing a similar allowedTopologies conversion for CSI

msau42 replied:
I'm not sure if multi-zonal PDs will be different.  This function is just creating a list of allowed zones.  The only difference is that for multi-zonal PDs, it will pick 2 zones from the list instead of 1.

ddebroy replied:
If we restrict the allowed set of keys in expressions for `allowedTopology` to only `LabelZoneFailureDomain` for now, it certainly simplifies things. The difference in logic compared to single-zone PD in this case will be: 

+ If both node and `allowedTopology` is present, we need to pick the node's zone + one separate zone from `allowedTopology` that is in the same region as node and return those two. 
+ If no node present, we can pick two zones from `allowedTopology` (as you said) and also need to make sure they are from the same region.

Sidenote: It does not appear if zones from multiple regions are specified in `zones` today, there are any checks [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util.go#L430) to pick zones only from the same region. If that is the case, I can enhance that.

msau42 replied:
I think we can split this out into two operations:
1. Convert allowed topologies to a list of zones
2. Figuring out how to select zones when node is set or not

Also, the zones are all within the same region.  We do not have cross-regional clusters today

msau42 replied:
@verult we may even want something like:
0. Flatten allowed topologies to a list of maps (csi topology format)
1. Convert list of maps to list of zones (and error if there's more than one key)
2. Figure out how to select zones from node + list of zones

msau42 replied:
> I don't think we'll be able to reuse the logic for CSI conversion as it'll be more complicated, for example there isn't a one-to-one mapping between allowedTopology and CSI topology representation.

Can you give an example of what you mean?  If there is no one-to-one mapping, then how will you be able to convert it?

verult replied:
> This could be a volume util that gce pd could use too
Should we pull it out into volume util when we get to the GCE plugin? It shouldn't be too hard to do it later on. It's easy to make the wrong assumptions about commonalities if this is done prematurely. I think this is more of a problem when creating abstraction boundaries so I don't have a strong preference either way for this case.

> 0. Flatten allowed topologies to a list of maps (csi topology format)
I don't think we should complicate the logic in in-tree volume plugins for the sake of aligning with CSI implementation details.

> Can you give an example of what you mean? If there is no one-to-one mapping, then how will you be able to convert it?
AllowedTopologies can have ORs of ANDs of ORs, so there are multiple possible representations of the same topology constraint. CSI only has ORs of ANDs. To convert from CSI representation to AllowedTopologies, we could pick a canonical representation, similar to how we deal with the same issue with volume node affinity.

msau42 replied:
Ok, let's consider refactoring this when we get to the CSI implementation.  But I feel the in-tree plugins use of allowed topologies is a subset of the CSI functionality, so it should be possible to get the list of zones out of the CSI representation.  Plus, here we're only talking about converting from AllowedTopologies -> CSI, not the other way around.

So for now, I think a common util that only handles zones should be fine

ddebroy replied:
> CSI only has ORs of ANDs

@verult Why does CSI not support ORs of ANDs of ORs? For multi-zonal implementations like GCE PD, the "last" ORs bit corresponding to the `values` is where the possible set of zones are specified:

```
allowedTopologies:
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-a
    - us-central1-b
    - us-central1-c
```
Will the above convention change for multi-zonal volumes in CSI?

verult replied:
AFAIU we could similarly have
```
allowedTopologies:
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-a
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-b
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-c
```
to represent the same thing (in both cases we are saying, these are the zones the volume plugin is free to choose from), and an equivalent form of this is how CSI represents it

msau42 commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
+							return nil, fmt.Errorf("multiple expressions with key %s cannot be specified in allowedTopology %v", exp.Key, term)
+						}
+						for _, value := range exp.Values {
+							allowedZones = append(allowedZones, value)
```
Should this be a set?  The same zone could have mistakenly been specified multiple times.

msau42 replied:
We can test with some examples.  But if it doesn't validate, we'll have to think about if we can strengthen it; the validation reuses much of the existing selector validation, and we can't tighten that.

verult replied:
> the validation reuses much of the existing selector validation, and we can't tighten that.

If it does turn out that current validation isn't sufficient, then I'd argue either the existing selector validation is somehow not strong enough, or maybe we shouldn't use these selectors

msau42 replied:
Can you open a bug for that? We can look into fixing that separately

verult replied:
Looked at the current validation code briefly and it doesn't seem like we have any structural validation today.

I'm happy to take this task if folks agree this is a good idea: https://github.com/kubernetes/kubernetes/issues/66184

msau42 commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
+			for _, term := range allowedTopologies {
+				zoneRequirementFound := false
+				for _, exp := range term.MatchLabelExpressions {
+					if exp.Key == kubeletapis.LabelZoneFailureDomain {
```
I wonder if this should error anything other than zone is specified.  Because that is information the user expects us to use, but we didn't.  cc @verult

ddebroy replied:
I guess they could be arbitrary labels. The example in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#multi-label-volume refers to `key: foo.io/rack` which should be okay?

msau42 replied:
Topology keys are a little different than label keys in Kubernetes, in that the topology keys are defined by the storage plugin and not the user, and the topology labels define the inherent restrictions of the storage provider.  So I think we shouldn't mix plugin and user-defined keys in allowedTopologies.  I think the safest course of action now is to error if there's a key we don't recognize.

msau42 commented on pkg/volume/aws_ebs/aws_util_test.go:
```diff
@@ -0,0 +1,232 @@
+			},
+			reject: true,
+		},
+		// fail since multiple zones specified in single allowedTopologies term DynamicProvisioningScheduling enabled
```
one more test case for single zone param + allowed topologies

verult replied:
What about nodeWithLabels + some allowedTopologies?

msau42 commented on pkg/volume/aws_ebs/aws_util_test.go:
```diff
@@ -0,0 +1,232 @@
+
+	for _, test := range tests {
+		if test.dynamicSchedulingEnabled {
+			utilfeature.DefaultFeatureGate.Set("DynamicProvisioningScheduling=true")
```
you probably need to disable the feature gate in the else case, to handle the times when you call continue

verult commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
-	}
-	glog.V(2).Infof("Successfully created EBS Disk volume %s", name)
+			// if node is set, we pick node's zone and ignore any allowedTopologies (since scheduler has accounted for it already)
+			if node.ObjectMeta.Labels == nil {
```
Is this check necessary? the `!ok` case below should be sufficient

verult commented on pkg/volume/aws_ebs/aws_util_test.go:
```diff
@@ -0,0 +1,232 @@
+	node                     *v1.Node
+	allowedTopologies        []v1.TopologySelectorTerm
+	dynamicSchedulingEnabled bool
+	reject                   bool
```
nit: Comment on what "reject" means

verult commented on pkg/volume/aws_ebs/aws_util_test.go:
```diff
@@ -0,0 +1,232 @@
+			},
+			reject: true,
+		},
+		// fail since multiple zones specified in single allowedTopologies term DynamicProvisioningScheduling enabled
```
"failed since *zones parameter and* multiple zones specified...

ddebroy commented:
/test pull-kubernetes-bazel-test

ddebroy commented:
/test pull-kubernetes-integration

ddebroy commented:
/test pull-kubernetes-e2e-gce-100-performance

ddebroy commented:
@msau42 @verult @jsafrane I have addressed the CR comments above and moved the common logic around zone population to `PopulateZoneConfig` in utils (so that GCE PD as well as Azure Disk can use it). PTAL.

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+	zones := make(sets.String)
+	for _, term := range allowedTopologies {
+		for _, exp := range term.MatchLabelExpressions {
+			if exp.Key == kubeletapis.LabelZoneFailureDomain {
```
We should document somewhere that only `LabelZoneFailureDomain` is supported on AWS (in a separate PR).

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+	if err != nil {
+		return nil, err
+	}
+	if allowedZones.Len() > 0 {
```
Should it return error if `allowedZones.Len() == 0 `? Such `allowedTopologies` are not very useful.

ddebroy replied:
Good point. I was not sure if https://github.com/kubernetes/kubernetes/issues/66184 will check for and reject `allowedZones.Len() == 0`. @verult let me know if we need to check for and reject `allowedZones.Len() == 0` here.

msau42 replied:
i'm not sure if api validation could handle this case.

you can have allowed topologies > 0 but not actually specify the zone label

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+		}
+		if allowedZones.Len() < replicas {
+			// for multi replica volumes, need as many allowedZones as replicas
+			return nil, fmt.Errorf("Require at least %d allowedZones for volumes with %d replicas", replicas, replicas)
```
errors start with lowercase

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+	zoneCfg.ReplicaZoneFromNodePresent = true
+
+	if replicas < 1 {
+		return nil, fmt.Errorf("Invalid number of replicas passed for Volume: %d", replicas)
```
errors start in lowercase

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+	}
+	if allowedZones.Len() > 0 {
+		if zonePresent || zonesPresent {
+			return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if allowedTopologies specified")
```
field names start in lowercase

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+					zones.Insert(value)
+				}
+			} else {
+				return nil, fmt.Errorf("Unsupported key found in MatchLabelExpression: %s", exp.Key)
```
errors start with lowercase

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+
+	// non-null node is specified from this point on
+	if zonePresent || zonesPresent {
+		return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode is set to WaitForFirstConsumer. Please specify AllowedTopology in StorageClass for constraining zones.")
```
fileld / object names start with lowercase

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+//    VolumeZoneConfig.ReplicaZoneFromNode, if present, should be used for one of the replica zones. Consider ZonesAlreadyPicked = 1
+//    VolumeZoneConfig.Zones, if present, should be used to choose N zones for the volume where N = replicas - ZonesAlreadyPicked
+//    if VolumeZoneConfig.Zones is absent, query and choose N zones from active zones where N = replicas - ZonesAlreadyPicked
+func PopulateZoneConfig(zonePresent, zonesPresent bool, zone, zones string, node *v1.Node, replicas int, allowedTopologies []v1.TopologySelectorTerm) (*VolumeZoneConfig, error) {
```
Nr. of replicas is hardcoded to 1, so lot of code below is useless. I'd suggest to focus on the topology in this PR and then send a subsequent one with multi-zonal volumes.

msau42 commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -86,68 +87,91 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
-	}
-
-	name, err := cloud.CreateDisk(volumeOptions)
+	zoneCfg, err := volumeutil.PopulateZoneConfig(zonePresent, zonesPresent, zone, zones, node, ebsMaxReplicasInAZ, allowedTopologies)
```
The aws cloud provider interface abstraction seems a bit off to me.  I feel like the aws cloud provider layer should not be processing the `zone` and `zones` arguments, which are StorageClass parameters.  It seems better suited for the Kubernetes plugin layer to process, and pass in the selected zone to the aws cloud provider layer.  Right now, both layers have to handle all the permutations of `zone` and `zones`.

If the aws cloud layer is refactored to push the zone argument processing up to the aws plugin, then I think this `PopulateZoneConfig` could be greatly simplified, and only needs to return a list of zones, either from allowedTopologies or from old `zone/zones`. @jsafrane wdyt?

ddebroy commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -554,14 +554,27 @@ func (c *awsElasticBlockStoreProvisioner) Provision(selectedNode *v1.Node, allow
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = requirements
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
```
Thanks for catching this - will fix. Indeed a case of a bad merge.

ddebroy commented:
Addressed CR comments so far with the following major changes:
1. Refactored the zone/zones parameter processing completely into the volume layer so that the cloud layer can just pass through a single zone to the ec2 APIs.
2. Removed extra logic (and tests) around handling of multi replicas. Will put that logic in a separate PR for GCE PD.

The Azure zone awareness PRs are also coming in now and that can use the common util in this PR for zonal volume creation.

ddebroy commented:
@jsafrane @msau42 @verult PTAL when you get a chance.

msau42 commented on pkg/volume/util/util_test.go:
```diff
@@ -1021,6 +1023,441 @@ func TestValidateZone(t *testing.T) {
+		Node                          *v1.Node
+		AllowedTopologies             []v1.TopologySelectorTerm
+		DynamicProvisioningScheduling bool
+		Reject                        bool // expect error due to validation failing
```
nit: add a comment when test args start, and when test expected results start

msau42 reviewed (commented):
/lgtm
just a minor nit

ddebroy commented:
/test pull-kubernetes-e2e-gce

msau42 commented:
/lgtm
@jsafrane can you also review, especially from aws perspective?

jsafrane commented:
AWS code is fine. Please squash the commits before it gets merged.

/approve
/lgtm cancel

jsafrane commented:
/approve no-issue

ddebroy committed:
Add DynamicProvisioningScheduling support for EBS

Signed-off-by: Deep Debroy <ddebroy@docker.com>

ddebroy commented:
/test pull-kubernetes-e2e-gce-100-performance

msau42 commented:
/lgtm

ddebroy commented:
/test pull-kubernetes-e2e-kops-aws
